doc/mgmt/SONiC_YANG_Model_Guidelines.md

```diff
@@ -0,0 +1,961 @@
+
+        description "PORTCHANNEL_INTERFACE part of config_db.json";
+
+        list PORTCHANNEL_INTERFACE_LIST {<<<<<
+        .....
+        .....
+        }
+}
+```
+
+### 19. In some cases it may be required to split an ABNF table into multiple YANG lists based on the data stored in the ABNF table. 
+
+Example : "INTERFACE" table stores VRF names to which an interface belongs, also it stores IP address of each interface. Hence it is needed to split them into two different YANG lists.
+
+#### ABNF
+```
+"INTERFACE" : {
+	"Ethernet1" : {
+		"vrf-name": "vrf1"
+	}
+	"Ethernet1|10.184.230.211/31": {
+	}
+}
+```
+#### YANG
+```
+......
+container sonic-interface {
+	container INTERFACE {
+		list INTERFACE_LIST {  // 1st list
+    			key ifname;
+			
+			leaf ifname {
+				type leafref { 
+				......
+				}
+			}
+			leaf vrf-name {
+				type leafref { 
+				......
+				}
+			}
+    			......
+		}
+
+		list INTERFACE_IPADDR_LIST { //2nd list
+   			key ifname, ip_addr;
+			
+  			leaf ifname {
+				type leafref { 
+				......
+				}
+			}
+  			leaf ip_addr {
+      				  type inet:ipv4-prefix;
+  			}
+			......
+		}
+	} 
+} 
+......
+```
+
+### 20. Add read-only nodes for state data using 'config false' statement. Define a separate top level container for state data. If state data is defined in other DB than CONFIG_DB, use extension 'sonic-ext:db-name' for defining the table present in other Redis DB. The default separator used in table key  is "|", if it is different, use 'sonic-ext:key-delim {separator};' YANG extension. This step applies when SONiC YANG is used as Northbound YANG.
+
+Example:
+
+#### YANG
+```
+container ACL_RULE {
+	list  ACL_RULE_LIST {
+
+	....
+	....
+		container state {
+			sonic-ext:db-name "APPL_DB"; //For example only
+			sonic-ext:key-delim ":";     //For example only
+			
+			config false;
+			description "State data";
+
+			leaf MATCHED_PACKETS {
+				type yang:counter64;
+			}
+
+			leaf MATCHED_OCTETS {
+				type yang:counter64;
+			}
+		}
+	}
+}
+```
+
+### 21. Define custom RPC for executing command like clear, reset etc. No configuration should change through such RPCs. Define 'input' and 'output' as needed, however they are optional. This step applies when SONiC YANG is used as Northbound YANG.
+
+Example:
+
+#### YANG
+```
+container sonic-acl {
+	....
+	....
+	rpc clear-stats {		
+		input {
+			leaf aclname {
+					type string;
+			}
+			
+			leaf rulename {
+				type string;
+			}
+		}
+	}
+}
+```
+
+### 22. Define Notification for sending out events generated in the system, e.g. link up/down or link failure event. This step applies  when SONiC YANG is used as Northbound YANG.
+
+Example:
+
+#### YANG
+```
+module sonic-port {
+	....
+	....
+	notification link_event {
+		leaf port {
+			type leafref {
+				path "../../PORT/PORT_LIST/ifname";
+			}
+		}
+	}
+}
+```
+
+### 23. Once YANG file is written, place it inside 'models/yang/' folder in 'sonic-mgmt-framework' framework repository.
```

Comment:
We should not mix it. For example, YANG Model can be part of even sonic-utilities Repo since we have "config-reload" command there. Config Reload must perform a validation as well. 
We may need a separate repo for YANG models or they should be part of sonic-utilities. YANG model can exist on switch even without any libraries. As mentioned even basic commands, such as config reload\load must be able to use it. Thanks.

Reply:
I will take out this point. The location of the YANG models needs to be discussed in Workgroup meeting.

Reply:
Done.